src/liballoc/string.rs

```diff
@@ -1848,6 +1848,21 @@ impl<'a, 'b> Pattern<'a> for &'b String {
     fn is_prefix_of(self, haystack: &'a str) -> bool {
         self[..].is_prefix_of(haystack)
     }
+
+    #[inline]
+    fn strip_prefix_from(self, haystack: &'a str) -> Option<&'a str> {
+        self[..].strip_prefix_from(haystack)
```

Comment:
So I'm having a little trouble following the flow of these methods. I'm going to botch some made up syntax and pretend `Module.method` is the `method` method on an instance of the `Module` struct but can you tell me if this what's going on?

We have
1. `String.strip_prefix_from` which calls `str.strip_prefix_from` which is an implementation of `Pattern.strip_prefix_from` optimized because we know the length of the `String` instance
1. `str.strip_prefix` which calls `prefix.strip_prefix_from(self)` which will call the optimized implementation if `prefix` is a `str` or `char`
1. `char.strip_prefix_from` which is an optimized implementation of `Pattern.strip_prefix_from`
1. a default implementation of `Pattern.strip_prefix_from` which is for ... I don't know 🤔 The PR description says "they have default implementations so the change is backwards compatible" - what does that mean? That sounds like, "if we no longer override `strip_prefix_from` for `str` or `char` they'll get the default implementation"? Or am I reading that wrong?

Reply:
Yep, exactly right! (4) is for anyone who wants to implement `Pattern` for their own type (say, a function) but doesn't want to go to the trouble of implementing a specialized `strip_prefix_from` or `strip_suffix_from method`—or perhaps their pattern is not of a fixed length and therefore there _isn't_ a way to optimize the method beyond constructing and using a full search.

Reply:
Ohh okay I understand now! I was originally confused because I didn't notice that, if there _wasn't_ a default implementation, implementors would _have_ to supply their own implementation. Forgot that optional methods without defaults aren't things :-) Thanks!

Reply:
You bet!